### PR TITLE
fuzz: improve utxo_snapshot target

### PR DESCRIPTION
**Pull request by mzumsande**
Add the possibility of giving more guidance to the creation of the metadata and/or coins, so that the fuzzer gets the chance
to reach more error conditions in ActivateSnapshot and sometimes successfully creates a valid snapshot.

This also changes the asserts for the success case that were outdated (after #29370) and only didn't result in a crash because the fuzzer wasn't able to reach this code before.

### CODE REVIEW

**Comment by fjahr** [comment]
Concept ACK

**Review by maflcko — APPROVED** [review]
lgtm. Thanks!

**Review by maflcko — COMMENTED** [review]
ACK b9ba1a73094f4ad593b527e23d2681f41c22539 🎯

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK b9ba1a73094f4ad593b527e23d2681f41c22539 🎯
iTX6VgeBMzgsveHFGIl5Qnsj/cWt0KYcv2B3dvhxmLm6ljOkWyT0bGPRv0zKVRDXMpcDBJcwelxlW8o6WHPEDw==
```

</details>

**Comment by TheCharlatan** [comment]
I checked if there is anything we could do in the testing setup to make the test faster. With this patchset, the test takes around 1:15 min for 10k runs on my machine. Removing the validation interface from the test setup cuts it down to 1:05 min , if I remove the networking setup, it takes 35 sec to complete.

I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?

**Comment by fjahr** [comment]
> I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?

That appears to be a bug. We actually require the mempool to be empty so it not being there should not be a problem. I will open a PR.

**Comment by fjahr** [comment]
> > I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?
> 
> That appears to be a bug. We actually require the mempool to be empty so it not being there should not be a problem. I will open a PR.

This should work https://github.com/fjahr/bitcoin/commit/21530b489eee6b8db4b9af45430d9d3cb43a459c, but on second thought I am not sure if we need this (yet). Conceptually it makes sense that we change this because, as I said, we don't need the mempool. However, since we don't use the mempool I think this should also not make an impact on the testing speed and even in blocksonly mode there still is a mempool, so this won't be an issue.

This might be getting off-topic so I can open a separate issue to move the discussion.

**Comment by fjahr** [comment]
utACK b9ba1a73094f4ad593b527e23d2681f41c225397

**Review by sedited — APPROVED** [review]
ACK b9ba1a73094f4ad593b527e23d2681f41c225397

... but it would be nice to address the nits, will re-ACK quickly :)

File: src/kernel/chainparams.cpp [hunk 1]
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
         };
 
         m_assumeutxo_data = {
-            {
+            {   // For use by unit tests
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/kernel/chainparams.cpp:498]: I didn't do that because there are already multiple files that use it (`validation_chainstatemanager_tests`, `validation_chainstate_tests`). Seems less likely that the same would happen for fuzz tests.

File: src/kernel/chainparams.cpp [hunk 2]
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
...
                 .hash_serialized = AssumeutxoHash{uint256S("0x6657b736d4fe4db0cbc796789e812d5dba7f5c143764b1b6905612f1830609d1")},
                 .nChainTx = 111,
                 .blockhash = uint256S("0x696e92821f65549c7ee134edceeeeaaa4105647a3c4fd9f298c0aec0ab50425c")
             },
+            {
+                // For use by fuzz target src/test/fuzz/utxo_snapshot.cpp
+                .height = 200,
+                .hash_serialized = AssumeutxoHash{uint256S("0x998b6a337899f04a1cd298ecf0fdc6ab66bfb428f00e5a48998505e91b3b55e0")},
+                .nChainTx = 201,
+                .blockhash = uint256S("0x5e93653318f294fb5aa339d00bbf8cf1c3515488ad99412c37608b139ea63b27,")
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/kernel/chainparams.cpp:509]: fixed, the "," went into the hash by mistake (and was ignored). I also fixed the `hash_serialized`, which was wrong somehow.

File: src/test/fuzz/utxo_snapshot.cpp [hunk 3]
```diff
@@ -41,8 +41,34 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
...
         AutoFile outfile{fsbridge::fopen(snapshot_path, "wb")};
-        const auto file_data{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
-        outfile << Span{file_data};
+        // Metadata
+        if (fuzzed_data_provider.ConsumeBool()) {
+            std::vector<uint8_t> metadata{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
+            outfile << Span{metadata};
+        } else {
+            DataStream data_stream{};
+            auto msg_start = Params().MessageStart();
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/test/fuzz/utxo_snapshot.cpp:50]: fixed, here and at one more spot.

File: src/test/fuzz/utxo_snapshot.cpp [hunk 4]
```diff
@@ -73,16 +99,18 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
...
         for (const auto& block : *g_chain) {
             Assert(coinscache.HaveCoin(COutPoint{block->vtx.at(0)->GetHash(), 0}));
             const auto* index{chainman.m_blockman.LookupBlockIndex(block->GetHash())};
-            const auto num_tx{Assert(index)->nTx};
-            Assert(num_tx == 1);
-            chain_tx += num_tx;
+            Assert(index);
+            Assert(index->nTx == 0);
+            if (index->nHeight == chainman.GetSnapshotBaseHeight()) {
+                Assert(index->nChainTx != 0);
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/test/fuzz/utxo_snapshot.cpp:108]: Done!

File: src/test/fuzz/utxo_snapshot.cpp [hunk 5]
```diff
@@ -41,8 +41,34 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
...
+        // Metadata
+        if (fuzzed_data_provider.ConsumeBool()) {
+            std::vector<uint8_t> metadata{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
+            outfile << Span{metadata};
+        } else {
+            DataStream data_stream{};
+            auto msg_start = Params().MessageStart();
+            int base_blockheight{fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 2 * COINBASE_MATURITY)};
+            uint256 base_blockhash{g_chain->at(base_blockheight - 1)->GetHash()};
+            uint64_t m_coins_count{fuzzed_data_provider.ConsumeIntegralInRange<uint64_t>(1, 2 * COINBASE_MATURITY)};
```
[the remark this answers is not in the document]
Comment (mzumsande) [reply src/test/fuzz/utxo_snapshot.cpp:53]: Done, use `*3`

**Comment by mzumsande** [comment]
Thanks for the reviews! I addressed the feedback by @maflcko and also corrected `hash_serialized` in the chainparams. Not sure why it was wrong before, I'm pretty sure I was able to activate the snapshot before opening the PR on another computer. Probably just an oversight, but maybe someone can confirm that everything is deterministic and the snapshot can be loaded now (e.g. by adding an assert and setting `base_blockheight` and `m_coins_count` to 200 it should crash almost immediately).

> I checked if there is anything we could do in the testing setup to make the test faster. (...)

Sound like good ideas for a follow-up!

**Review by sedited — APPROVED** [review]
ACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

**Comment by fjahr** [comment]
utACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

**Comment by mzumsande** [comment]
> maybe someone can confirm that everything is deterministic and the snapshot can be loaded now (e.g. by adding an assert and setting `base_blockheight` and `m_coins_count` to 200 it should crash almost immediately).

FYI, confirmed for myself on another computer that this was just an error, and the snapshot can be loaded by the fuzzer.

File: src/kernel/chainparams.cpp [hunk 6]
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
...
                 .hash_serialized = AssumeutxoHash{uint256S("0x6657b736d4fe4db0cbc796789e812d5dba7f5c143764b1b6905612f1830609d1")},
                 .nChainTx = 111,
                 .blockhash = uint256S("0x696e92821f65549c7ee134edceeeeaaa4105647a3c4fd9f298c0aec0ab50425c")
             },
+            {
+                // For use by fuzz target src/test/fuzz/utxo_snapshot.cpp
+                .height = 200,
+                .hash_serialized = AssumeutxoHash{uint256S("0x4f34d431c3e482f6b0d67b64609ece3964dc8d7976d02ac68dd7c9c1421738f2")},
+                .nChainTx = 201,
+                .blockhash = uint256S("0x5e93653318f294fb5aa339d00bbf8cf1c3515488ad99412c37608b139ea63b27"),
```
Comment (ryanofsky) [inline src/kernel/chainparams.cpp:509]: In commit "fuzz: improve utxo_snapshot target" (de71d4dece604907afc4fc26b7788e9c1a4cbecb)

Since fuzz test is not deterministic, and doesn't give us an easy way to know hash values here actually allow snapshot to be loaded, it might be good to write a separate regtest generating a an empty chain with height 200 and verifying these hashes.

**Review by ryanofsky — COMMENTED** [review]
Post-merge code review ACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

This seems like a good change in principle, but I'm not actually sure what additional test coverage it provides. The only valid snapshot that can be loaded seems to be one that is hardcoded for a chain with height 200 and no transactions except for coinbases. After the snapshot is generated there is only a single `ConsumeBool` call determining whether or not headers for the blocks are added before loading the snapshot.

So basically it creates a hardcoded snapshot, and then tests loading it with and without headers. I could be missing something, but it seems like that is not really taking advantage of the fuzzer or testing something that would be not more convenient to run as a unit test.

It would probably open up a lot more flexibility for the fuzz test if it were able to modify the `m_assumeutxo_data` vector instead of relying on hardcoded values.

**Comment by mzumsande** [comment]
> This seems like a good change in principle, but I'm not actually sure what additional test coverage it provides. The only valid snapshot that can be loaded seems to be one that is hardcoded for a chain with height 200 and no transactions except for coinbases. After the snapshot is generated there is only a single `ConsumeBool` call determining whether or not headers for the blocks are added before loading the snapshot.

Note that loading a valid snapshot is not the only goal of this PR. It also allows for a better coverage of `ActivateSnapshot()` which was previously hard to reach because the fuzzer already had a hard time fuzzing valid metadata (after #29612) and would abort without it. E.g. it has zero coverage in https://maflcko.github.io/b-c-cov/fuzz.coverage/src/validation.cpp.gcov.html though I'm not sure how up-do-date the corpus is, and fuzzing locally I also didn't reach `ActivateSnapshot()` (maybe I wasn't patient enough).

I think for more interesting fuzzing we'd have to think about allowing dynamic registration of snapshots in tests (or enabling an option to just skip the check for the hash in chainparams). 

Some time ago, I thought about this in the context of functional tests (to avoid having to edit chainparams whenever we change something in a functional test), but after realizing that the registered snapshot data from chainparams is not only necessary during registration but also in a restart (`nChainTx`) it seemed too crude to me and I didn't open a PR. Might make more sense for fuzzing though.

**Review by Theghost256 — APPROVED** [review]
